### ISSUE TITLE
Fix S3267 FP/FN: ReadOnlySpan / Span in combination with logical pattern results in false positive

### ISSUE THREAD

**Issue by linkdotnet** [SonarSource/sonar-dotnet#7730]
### Description

When enumerating over a `Span` / `ReadOnlySpan` rule S3267 alerts with:
> Error S3267 : Loops should be simplified with "LINQ" expressions (https://rules.sonarsource.com/csharp/RSPEC-3267)

### Repro steps

```csharp
private static int GetWordCount(ReadOnlySpan<char> content)
{
    var wordCount = 0;
    foreach (var c in content)
    {
        if (c is ' ' or '\n' or '\r')
        {
            wordCount++;
        }
    }

    return wordCount;
}
```

The following code will *not* raise the error:
```csharp
private static int GetWordCount(ReadOnlySpan<char> content)
{
    var wordCount = 0;
    foreach (var c in content)
    {
        if (c == ' ' || c == '\n' || c == '\r')
        {
            wordCount++;
        }
    }

    return wordCount;
}
```

### Expected behavior

No warning/error.

### Actual behavior

Warning/error.

### Related information

* C#/VB.NET Plugins version: 9.7
* Visual Studio version
* MSBuild / dotnet version: 8.0 preview 6
* SonarScanner for .NET version (if used)
* Operating System: MacOS

**Comment by antonioaversa** [comment]
Hi @linkdotnet, thanks for reporting this issue.

I can confirm that you have identified a False Positive case: the rule [S3267](https://sonarsource.github.io/rspec/#/rspec/S3267/csharp) doesn't make sure that the object being iterated is an `IEnumerable`, or an `IQueryable`, so the code fix proposes a change that is incorrect.

As you also pointed out, the behavior of the rule is inconsistent with respect to logical pattern matching and logical operators: while the first is recognized as a valid condition for the rule and the fix, the second is not. It's basically a limitation of the current implementation of the rule.

I have added a reproducer for all the cases I could identify in [this PR](https://github.com/SonarSource/sonar-dotnet/pull/7766).
Waiting for us to provide a fix for the issue, which requires extending the current implementation, please mark the issue as False Positive in SonarQube/SonarCloud or use the standard suppression mechanism or the `.editorconfig` from Visual Studio. 
If the rule is particularly noisy in your codebase, you can also disable it entirely.

Best regards,
Antonio

**Comment by pavel-mikula-sonarsource** [comment]
Internal ticket [NET-1706](https://sonarsource.atlassian.net/browse/NET-1706)

[NET-1706]: https://sonarsource.atlassian.net/browse/NET-1706?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

**Comment by Tim-Pohlmann** [comment]
This has been fixed in [10.12](https://github.com/SonarSource/sonar-dotnet/releases/tag/10.12.0.118525)

**Comment by burnchar** [comment]
This situation appears to be a similar false positive. Using version 10.15.0.120848:

```
	public static bool EndsWithAny(this ReadOnlySpan<char> value, IEnumerable<string> enders, StringComparison stringComparison = StringComparison.Ordinal)
	{
		foreach(var str in enders)
			if(value.EndsWith(str, stringComparison))
				return true;
		return false;
	}
```

**Comment by Tim-Pohlmann** [comment]
@burnchar That looks like a true positive to me. `enders` is an `IEnumerable<T>` and thus the snippet can be rewritten to:
```csharp
public static bool EndsWithAny(this ReadOnlySpan<char> value, IEnumerable<string> enders, StringComparison stringComparison = StringComparison.Ordinal) =>
    enders.Any(x => value.EndsWith(x, stringComparison));
```

**Comment by burnchar** [comment]
@Tim-Pohlmann  Good point. I did not realize that the analyzer's coverage was so broad. Thank you.
